SampleApplications/SDK/Opc.Ua.Client.ComplexTypes/ComplexTypeSystem.cs

```diff
@@ -0,0 +1,792 @@
+        }
+
+        /// <summary>
+        /// Browse for the type and encoding id for a dictionary component.
+        /// </summary>
+        /// <remarks>
+        /// To find the typeId and encodingId for a dictionary type definition:
+        /// i) inverse browse the description to get the encodingid
+        /// ii) from the description inverse browse for encoding 
+        /// to get the subtype typeid 
+        /// iii) load the DataType node 
+        /// </remarks>
+        /// <param name="nodeId"></param>
+        /// <param name="typeId"></param>
+        /// <param name="encodingId"></param>
+        /// <returns></returns>
+        private bool BrowseTypeIdsForDictionaryComponent(
+            NodeId nodeId,
+            out ExpandedNodeId typeId,
+            out ExpandedNodeId encodingId,
+            out DataTypeNode dataTypeNode)
+        {
+            typeId = ExpandedNodeId.Null;
+            encodingId = ExpandedNodeId.Null;
+            dataTypeNode = null;
+
+            var references = m_session.NodeCache.FindReferences(
+                nodeId,
+                ReferenceTypeIds.HasDescription,
+                true,
+                false
+                );
+
+            if (references.Count == 1)
+            {
+                encodingId = references.First().NodeId;
+                references = m_session.NodeCache.FindReferences(
+                    encodingId,
+                    ReferenceTypeIds.HasEncoding,
+                    true,
+                    false
+                    );
+                encodingId = NormalizeExpandedNodeId(encodingId);
+
+                if (references.Count == 1)
+                {
+                    typeId = references.First().NodeId;
+                    dataTypeNode = m_session.NodeCache.Find(typeId) as DataTypeNode;
+                    typeId = NormalizeExpandedNodeId(typeId);
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Browse for the property.
+        /// </summary>
+        /// <remarks>
+        /// Browse for property (type description) of an enum datatype.
+        /// </remarks>
+        /// <param name="nodeId"></param>
+        /// <returns></returns>
+        private INode BrowseForSingleProperty(
+            ExpandedNodeId nodeId)
+        {
+            var references = m_session.NodeCache.FindReferences(
+                nodeId,
+                ReferenceTypeIds.HasProperty,
+                false,
+                false
+                );
+            return references.FirstOrDefault();
+        }
+
+        /// <summary>
+        /// Load all subTypes and optionally nested subtypes of a type definition.
+        /// Filter for all subtypes or only subtypesoutside the default namespace.
+        /// </summary>
+        private IList<INode> LoadDataTypes(ExpandedNodeId dataType, bool nestedSubTypes = false, bool filterUATypes = true)
```

Comment:
Would be cool to have an async or Begin/End version of it also - eventually....

Reply:
It is private so it can be improved later...